lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,653 @@
+  LLVM_DEBUG(llvm::dbgs() << "Instance graph:\n");
+  for (auto it : moduleInstances) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "- " << it.first->getName() << " "
+               << it.first->getAttrOfType<StringAttr>("sym_name") << "\n");
+    for (auto inst : it.second)
+      LLVM_DEBUG(llvm::dbgs()
+                 << "  - " << inst.nameAttr() << " in \""
+                 << inst->getParentOfType<FModuleOp>().getName() << "\"\n");
+  }
+#endif
+}
+
+ArrayRef<InstancePath> InstanceGraph::getAbsolutePaths(Operation *op) {
+  if (!isa<FModuleOp, FExtModuleOp>(op))
+    return {};
+
+  // If we have reached the circuit root, we're done.
+  if (mainModule == op) {
+    static InstancePath empty{};
+    return empty; // array with single empty path
+  }
+
+  // Otherwise look up the instances of this module.
+  auto it = moduleInstances.find(op);
+  if (it == moduleInstances.end())
+    return {};
+
+  // Fast path: hit the cache.
+  auto cached = absolutePathsCache.find(op);
+  if (cached != absolutePathsCache.end())
+    return cached->second;
+
+  // For each instance, collect the instance paths to its parent and append the
+  // instance itself to each.
+  SmallVector<InstancePath, 8> extendedPaths;
+  for (auto inst : it->second) {
+    auto instPaths = getAbsolutePaths(inst->getParentOfType<FModuleOp>());
+    extendedPaths.reserve(instPaths.size());
+    for (auto path : instPaths) {
+      extendedPaths.push_back(appendInstance(path, inst));
+    }
+  }
+
+  // Move the list of paths into the bump allocator for later quick retrieval.
+  ArrayRef<InstancePath> pathList;
+  if (!extendedPaths.empty()) {
+    auto paths = allocator.Allocate<InstancePath>(extendedPaths.size());
+    std::copy(extendedPaths.begin(), extendedPaths.end(), paths);
+    pathList = ArrayRef<InstancePath>(paths, extendedPaths.size());
+  }
+
+  absolutePathsCache.insert({op, pathList});
+  return pathList;
+}
+
+InstancePath InstanceGraph::appendInstance(InstancePath path, InstanceOp inst) {
+  size_t n = path.size() + 1;
+  auto newPath = allocator.Allocate<InstanceOp>(n);
+  std::copy(path.begin(), path.end(), newPath);
+  newPath[path.size()] = inst;
+  return InstancePath(newPath, n);
+}
+
+/// Return a version of `path` that skips all front instances it has in common
+/// with `other`.
+static InstancePath stripCommonPrefix(InstancePath path, InstancePath other) {
+  while (!path.empty() && !other.empty() && path.front() == other.front()) {
+    path = path.drop_front();
+    other = other.drop_front();
+  }
+  return path;
+}
+
+//===----------------------------------------------------------------------===//
+// Static information gathered once upfront
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Attributes used throughout the annotations.
+struct Strings {
+  MLIRContext *const context;
+  Strings(MLIRContext *context) : context(context) {}
+
+  StringAttr dataTapsClass = StringAttr::get(
```

Comment:
I'd recommend making these just be `const char *'s`.  These are not going to be common, and the string compares will be fast because the lengths are checked first (these are long strings).  I'd default to simplest possible code unless there is a reason to optimize it.

This will also make the pass slightly faster for files that don't use GC.

Reply:
That makes sense. Changing these to `const char *` also allows for them to be `static constexpr`s, and the struct itself is not needed to be passed around by value.